File: src/database/firebase/collections/serviceCollection.js
import { db, COLLECTIONS } from '../config';
import { 
  collection, 
  doc,
  addDoc, 
  getDoc,
  getDocs, 
  updateDoc,
  deleteDoc,
  query, 
  where, 
  orderBy,
  onSnapshot,
  increment,
  serverTimestamp 
} from 'firebase/firestore';

const servicesCollection = collection(db, COLLECTIONS.SERVICES);

/**
 * ADD SERVICE
 */
export const addService = async (serviceData) => {
  try {
    const docRef = await addDoc(servicesCollection, {
      ...serviceData,
      createdAt: serverTimestamp(),
      updatedAt: serverTimestamp()
    });
    return { success: true, serviceId: docRef.id };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * GET SERVICE BY ID
 */
export const getService = async (serviceId) => {
  try {
    const serviceDoc = await getDoc(doc(servicesCollection, serviceId));
    if (serviceDoc.exists()) {
      return { success: true, data: { id: serviceDoc.id, ...serviceDoc.data() } };
    } else {
      return { success: false, error: 'Service not found' };
    }
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * GET ALL SERVICES (Real-time)
 */
export const subscribeToAllServices = (callback) => {
  const q = query(
    servicesCollection,
    where('isActive', '==', true),
    orderBy('createdAt', 'desc')
  );
  
  return onSnapshot(q, (snapshot) => {
    const services = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    }));
    callback(services);
  });
};

/**
 * GET SERVICES BY VENDOR (Real-time)
 */
export const subscribeToVendorServices = (vendorId, callback) => {
  const q = query(
    servicesCollection,
    where('vendorId', '==', vendorId),
    orderBy('createdAt', 'desc')
  );
  
  return onSnapshot(q, (snapshot) => {
    const services = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    }));
    callback(services);
  });
};

/**
 * GET SERVICES BY CATEGORY (Real-time)
 */
export const subscribeToServicesByCategory = (category, callback) => {
  const q = query(
    servicesCollection,
    where('category', '==', category),
    where('isActive', '==', true),
    orderBy('createdAt', 'desc')
  );
  
  return onSnapshot(q, (snapshot) => {
    const services = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    }));
    callback(services);
  });
};

/**
 * UPDATE SERVICE
 */
export const updateService = async (serviceId, updates) => {
  try {
    await updateDoc(doc(servicesCollection, serviceId), {
      ...updates,
      updatedAt: serverTimestamp()
    });
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * DELETE SERVICE
 */
export const deleteService = async (serviceId) => {
  try {
    await deleteDoc(doc(servicesCollection, serviceId));
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * INCREMENT TOTAL BOOKINGS
 */
export const incrementTotalBookings = async (serviceId) => {
  try {
    await updateDoc(doc(servicesCollection, serviceId), {
      totalBookings: increment(1)
    });
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

File: app/src/main/java/com/example/bunnix/database/firebase/collections/ServiceCollection.kt
package com.example.bunnix.database.firebase.collections

import com.example.bunnix.database.config.FirebaseConfig
import com.example.bunnix.database.models.Service
import com.google.firebase.firestore.Query
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

object ServiceCollection {

    private val collection = FirebaseConfig.firestore.collection(FirebaseConfig.Collections.SERVICES)

    // GET ALL SERVICES
    fun getAllServices(): Flow<List<Service>> = callbackFlow {
        val listener = collection
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val services = snapshot?.toObjects(Service::class.java) ?: emptyList()
                trySend(services)
            }
        awaitClose { listener.remove() }
    }

    // GET SERVICES BY VENDOR
    fun getServicesByVendor(vendorId: String): Flow<List<Service>> = callbackFlow {
        val listener = collection
            .whereEqualTo("vendorId", vendorId)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val services = snapshot?.toObjects(Service::class.java) ?: emptyList()
                trySend(services)
            }
        awaitClose { listener.remove() }
    }

    // ADD SERVICE
    suspend fun addService(service: Service): Result<String> {
        return try {
            val docRef = collection.add(service).await()
            Result.success(docRef.id)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // UPDATE SERVICE
    suspend fun updateService(serviceId: String, updates: Map<String, Any>): Result<Unit> {
        return try {
            collection.document(serviceId).update(updates).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // DELETE SERVICE
    suspend fun deleteService(serviceId: String): Result<Unit> {
        return try {
            collection.document(serviceId).delete().await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}